src/uct/ib/ud/accel/ud_mlx5.c

```diff
@@ -716,7 +716,7 @@ static ucs_status_t uct_ud_mlx5_iface_create_qp(uct_ib_iface_t *ib_iface,
     uct_ib_mlx5_qp_t *qp = &iface->tx.wq.super;
     ucs_status_t status;
 
-    status = uct_ib_mlx5_iface_create_qp(ib_iface, qp, attr);
+    status = uct_ib_mlx5_iface_create_qp(ib_iface, qp, (uct_ib_mlx5_qp_attr_t *)attr);
```

Comment:
why need to cast? can we pass uct_ib_mlx5_qp_attr_t * to this function?

Reply:
because ud_iface_ops.create_qp is callback common to verbs and mlx5

Reply:
can we use ucs_derived_of ?

Reply:
We can, but I would prefer allocate new uct_ib_mlx5_qp_attr_t in this function and initialize it with upper one.
```
diff --git a/src/uct/ib/ud/accel/ud_mlx5.c b/src/uct/ib/ud/accel/ud_mlx5.c
index 2528fef..6769f3d 100644
--- a/src/uct/ib/ud/accel/ud_mlx5.c
+++ b/src/uct/ib/ud/accel/ud_mlx5.c
@@ -709,14 +709,17 @@ static void uct_ud_mlx5_iface_event_cq(uct_ib_iface_t *ib_iface,
 }

 static ucs_status_t uct_ud_mlx5_iface_create_qp(uct_ib_iface_t *ib_iface,
-                                                uct_ib_qp_attr_t *attr,
+                                                uct_ib_qp_attr_t *ib_attr,
                                                 struct ibv_qp **qp_p)
 {
     uct_ud_mlx5_iface_t *iface = ucs_derived_of(ib_iface, uct_ud_mlx5_iface_t);
     uct_ib_mlx5_qp_t *qp = &iface->tx.wq.super;
+    uct_ib_mlx5_qp_attr_t attr = {};
     ucs_status_t status;

-    status = uct_ib_mlx5_iface_create_qp(ib_iface, qp, (uct_ib_mlx5_qp_attr_t *)attr);
+    attr.super = *ib_attr;
+
+    status = uct_ib_mlx5_iface_create_qp(ib_iface, qp, &attr);
     if (status != UCS_OK) {
         return status;
     }
```

Reply:
in this suggestion, where is the mlx5 part (mmio mode) initialized?

Reply:
it's not
UD has no DEVX flow
mmio set [later](https://github.com/openucx/ucx/blob/master/src/uct/ib/ud/accel/ud_mlx5.c#L818) in iface constructor

Reply:
better not pass uninitialized struct to a function, it's recipe for unexpected bugs
if a function does not need a parameter - it should not be passed at all

Reply:
my point was to pass it initialized to zero

Reply:
but what is the meaning of mmio_mode=0? is it really what we want as default?

Reply:
indeed. let's set to UCT_IB_MLX5_MMIO_MODE_LAST

Reply:
and assert that mode!= LAST when it's used?